### PR TITLE
A little better RegExps and refactor

### PR DESCRIPTION
**Pull request by thepheer**
Wanted to improve one little regexp but ended up with a few more changes. :)

### CODE REVIEW

**Comment by fatfisz** [comment]
All looks good, but I'm not too sure about "\r" in there. In many places in the code only "\n" is used, so I'd leave it like that for consistency - otherwise a bigger refactor would be in order.

**Comment by thepheer** [comment]
A am also unsure about that. I personaly use only `\n` in my code. But I think it would be a good idea to keep `\r`'s because changing the style of newlines is not what people could expect from `stripIndents`.

Also, as far as I can tell from the code, `\r`'s were untouched before 900aa00a952d67a5295c387e0759da2448931c51 . Now `stripIndents` removes them but `stripIndent` keeps. In fact it removes them in a weird way: sequences of newlines like `\r\n\r\n\r\n` became `\n\n\n` but regular strings like `foo\r\nbar` keep that `\r` in the end. So `stripIndents` just mixes styles of newlines, which is not really a good thing.

I think a decent solution would be not to mess with `\r`'s at all and a better (?) one would be removing them only when they're being a part of identation (for some really strange reason, haha) which requires some more complex regexps.

I also consider removing `\r`'s completely and not really against it. At least it won't affect me and my code. But again, we need something better than `[^\S\n]` since it just does weird things.

What do you think? :)

**Comment by fatfisz** [comment]
I'd leave `[^\S\n]` and wait until people complain 😉 If push comes to shove I think having a "turn `\n` into `\r\n`" tag and a "turn `\r\n` into `\n`" tag will be a nice better solution. This, or a big refactor for the whole lib - I think it's better to **consistently not work** with something than **work in some special cases**.

Btw. the behavior was not changed in https://github.com/declandewet/common-tags/commit/900aa00a952d67a5295c387e0759da2448931c51, because previously splitting was done by `\n`, and then `'\r'.trimLeft() === ''` (at least where `String.prototype.trimLeft` is implemented 😄).

**Comment by thepheer** [comment]
Ah, now I see. I saw `trimLeft` and it was taken literally by my brain. 😄

I thought I'm returning consistensy back with my PR (at least for `stripIndents`), but in reality I'm doing the opposite. I had to test more before doing PR, sorry about that.

Well, since my PR doesn't really make sence anymore, should I close it?

**Comment by fatfisz** [comment]
I really dig early returns instead of reassigning to arguments, so I'd merge that happily :)

File: src/stripIndentTransformer/stripIndentTransformer.js [hunk 1]
```diff
@@ -9,21 +9,19 @@ const stripIndentTransformer = (type = 'initial') => ({
...
       // remove the shortest leading indentation from each line
-      const match = endResult.match(/^[ \t]*(?=\S)/gm)
-      // return early if there's nothing to strip
-      if (match === null) {
-        return endResult
+      const match = endResult.match(/^[^\S\r\n]*(?=\S)/gm)
+      const indent = match && Math.min(...match.map(el => el.length))
+      if (indent) {
+        const regexp = new RegExp(`^.{${indent}}`, 'gm')
+        endResult = endResult.replace(regexp, '')
```
Comment (fatfisz) [inline src/stripIndentTransformer/stripIndentTransformer.js]: You can return early here too.

Comment (thepheer) [reply src/stripIndentTransformer/stripIndentTransformer.js]: Sure thing!

File: src/stripIndentTransformer/stripIndentTransformer.js [hunk 2]
```diff
@@ -9,21 +9,19 @@ const stripIndentTransformer = (type = 'initial') => ({
   onEndResult (endResult) {
     if (type === 'initial') {
       // remove the shortest leading indentation from each line
-      const match = endResult.match(/^[ \t]*(?=\S)/gm)
-      // return early if there's nothing to strip
-      if (match === null) {
-        return endResult
+      const match = endResult.match(/^[^\S\r\n]*(?=\S)/gm)
```
Comment (fatfisz) [inline src/stripIndentTransformer/stripIndentTransformer.js]: Please remove `\r` here.

File: src/stripIndentTransformer/stripIndentTransformer.js [hunk 3]
```diff
@@ -9,21 +9,19 @@ const stripIndentTransformer = (type = 'initial') => ({
...
-      endResult = indent > 0 ? endResult.replace(regexp, '') : endResult
-    } else if (type === 'all') {
+      return endResult
+    }
+    if (type === 'all') {
       // remove all indentation from each line
-      endResult = endResult.replace(/(?:\n[^\S\n]*)/g, '\n')
-    } else {
-      throw new Error(`Unknown type: ${type}`)
+      return endResult.replace(/^[^\S\r\n]+/gm, '')
```
Comment (fatfisz) [inline src/stripIndentTransformer/stripIndentTransformer.js]: And here too

**Comment by thepheer** [comment]
This PR is definitely more trouble than it's worth now. 😄

**Comment by fatfisz** [comment]
Really sorry about this, it was not my intention to drag this out for so long. If you haven't done that refactor, I'd probably do the same thing eventually, but since you put in some work I didn't want to dismiss it.

Let me know here if you want to be added to the list of the contributors; I'll add you in a separate commit then so to not block this PR anymore.

**Comment by thepheer** [comment]
I mean I should have changed `(/(?:\n[^\S\n]*)/g, '\n')` to `(/^[^\S\n]+/gm, '')` just as I planned it when making a fork. But I did more than needed without enough testing and thinkng about consequences. So I am the only troublemaker here and I should say sorry about this, not you. 😄

I though github manages contributors automatically?

**Pull request merged**